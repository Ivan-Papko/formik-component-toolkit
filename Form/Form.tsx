import { Formik } from 'formik'

import { FormikMultiSelect } from './FormikMultiSelect/FormikMultiSelect'
import { FormikAutocomplete } from './FormikAutocomplete'
import { FormikCheckbox } from './FormikCheckbox'
import { FormikCopyInput } from './FormikCopyInput'
import { FormikCountryField } from './FormikCountryField'
import { FormikDatePicker } from './FormikDatePicker'
import { FormikForm } from './FormikForm'
import { FormikPasswordField } from './FormikPasswordField'
import { FormikRegionField } from './FormikRegionField'
import { FormikSelect } from './FormikSelect'
import { FormikSubmit } from './FormikSubmit'
import { FormikSwitch } from './FormikSwitch'
import {
  type IFormikTextFieldProps,
  FormikTextField,
  FormikTextFieldWithRef,
} from './FormikTextField'
import { FormikTimePicker } from './FormikTimePicker'
import { FormikValidatingInput } from './FormikValidatingInput'
import { FormikWrapper } from './FormikWrapper'

/**
 * @deprecated Use `HookForm` instead.
 */
const Form = {
  Context: Formik,
  Form: FormikForm,
  Wrapper: FormikWrapper,
  TextField: FormikTextField,
  TextFieldWithRef: FormikTextFieldWithRef,
  PasswordField: FormikPasswordField,
  Checkbox: FormikCheckbox,
  ValidatingInput: FormikValidatingInput,
  CountryField: FormikCountryField,
  /**
   * @deprecated
   */
  RegionField: FormikRegionField,
  Submit: FormikSubmit,
  Switch: FormikSwitch,
  MultiSelect: FormikMultiSelect,
  Select: FormikSelect,
  Autocomplete: FormikAutocomplete,
  CopyInput: FormikCopyInput,
  TimePicker: FormikTimePicker,
  DatePicker: FormikDatePicker,
}

export { Form, type IFormikTextFieldProps }
export default Form
