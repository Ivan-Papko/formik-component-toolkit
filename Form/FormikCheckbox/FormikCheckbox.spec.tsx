import { useField } from 'formik'

import { render } from '@cloudbees/shared-test-utils/render'

import { Form } from '../Form'

// Mock the useField hook from Formik
jest.mock('formik', () => ({
  ...jest.requireActual('formik'),
  useField: jest.fn(),
}))

const factory = (onChange?: () => void) =>
  render(
    <Form.Context
      initialValues={{}}
      onSubmit={() => {
        return
      }}
    >
      <Form.Form>
        <Form.Wrapper>
          <Form.Checkbox name="name" label="some label" onChange={onChange} />
        </Form.Wrapper>
      </Form.Form>
    </Form.Context>,
  )

describe('Checkbox', () => {
  beforeEach(() => {
    ;(useField as jest.Mock).mockImplementation(() => [
      { onChange: jest.fn() },
      { touched: false, error: null },
    ])
  })
  it('should render a checkbox component', () => {
    const { baseElement } = factory()
    expect(baseElement).toMatchSnapshot()
  })

  it('should display the correct label', () => {
    const { screen } = factory()
    expect(screen.getByLabelText('some label')).toBeInTheDocument()
  })

  it('should toggle checkbox state on click', async () => {
    const { setup, screen } = factory()
    const user = setup()
    const checkbox = screen.getByRole('checkbox')
    expect(checkbox).not.toBeChecked()

    await user.click(checkbox)
    expect(checkbox).toBeChecked()

    await user.click(checkbox)
    expect(checkbox).not.toBeChecked()
  })

  it('should call onChange handler when checkbox is clicked', async () => {
    const handleChange = jest.fn()
    const { screen, setup } = factory(handleChange)
    const user = setup()

    const checkbox = screen.getByRole('checkbox')
    await user.click(checkbox)
    expect(handleChange).toHaveBeenCalledTimes(1)
  })

  it('should not be crashed in onChangeHandler is not passed', async () => {
    const { setup, screen } = factory()
    const user = setup()

    const checkbox = screen.getByRole('checkbox')
    await user.click(checkbox)
  })

  it('should show error message when there is an error', () => {
    ;(useField as jest.Mock).mockImplementation(() => [
      { onChange: jest.fn() },
      { touched: true, error: 'Error message' },
    ])
    const { screen } = factory()
    expect(screen.getByText('Error message')).toBeInTheDocument()
  })
})
