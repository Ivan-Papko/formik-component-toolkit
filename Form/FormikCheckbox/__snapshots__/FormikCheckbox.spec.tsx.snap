// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should render a checkbox component 1`] = `
.emotion-0 {
  margin-top: 24px;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 16px;
  width: 100%;
}

.emotion-1 {
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.emotion-2 {
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  vertical-align: middle;
  -webkit-tap-highlight-color: transparent;
  margin-left: -11px;
  margin-right: 16px;
}

.emotion-2.Mui-disabled {
  cursor: default;
}

.emotion-2 .MuiFormControlLabel-label.Mui-disabled {
  color: rgba(0, 0, 0, 0.38);
}

.emotion-3 {
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 50%;
  padding: 9px;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: rgba(0, 0, 0, 0.6);
}

.emotion-3::-moz-focus-inner {
  border-style: none;
}

.emotion-3.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-3 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-3:hover {
  background-color: rgba(25, 118, 210, 0.04);
}

@media (hover: none) {
  .emotion-3:hover {
    background-color: transparent;
  }
}

.emotion-3.Mui-checked,
.emotion-3.MuiCheckbox-indeterminate {
  color: #1976d2;
}

.emotion-3.Mui-disabled {
  color: rgba(0, 0, 0, 0.26);
}

.emotion-4 {
  cursor: inherit;
  position: absolute;
  opacity: 0;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  margin: 0;
  padding: 0;
  z-index: 1;
}

.emotion-5 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
}

.emotion-6 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

.emotion-7 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  letter-spacing: 0.00938em;
}

<body>
  <div>
    <form
      action="#"
    >
      <div
        class="MuiBox-root emotion-0"
      >
        <div
          class="MuiFormGroup-root emotion-1"
        >
          <label
            class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd emotion-2"
          >
            <span
              class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium emotion-3"
              label="some label"
            >
              <input
                class="PrivateSwitchBase-input emotion-4"
                data-indeterminate="false"
                name="name"
                type="checkbox"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-5"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root emotion-6"
              />
            </span>
            <span
              class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label emotion-7"
            >
              some label
            </span>
          </label>
        </div>
      </div>
    </form>
  </div>
</body>
`;
